src/Providers/OpenAiCompatibleImplementation/AbstractOpenAiCompatibleTextGenerationModel.php

```diff
@@ -0,0 +1,727 @@
+        if (is_array($functionDeclarations)) {
+            $params['tools'] = $this->prepareToolsParam($functionDeclarations);
+        }
+
+        $outputMimeType = $config->getOutputMimeType();
+        if ('application/json' === $outputMimeType) {
+            $outputSchema = $config->getOutputSchema();
+            $params['response_format'] = $this->prepareResponseFormatParam($outputSchema);
+        }
+
+        /*
+         * Any custom options are added to the parameters as well.
+         * This allows developers to pass other options that may be more niche or not yet supported by the SDK.
+         */
+        $customOptions = $config->getCustomOptions();
+        foreach ($customOptions as $key => $value) {
+            if (isset($params[$key])) {
+                throw new InvalidArgumentException(
+                    sprintf(
+                        'The custom option "%s" conflicts with an existing parameter.',
+                        $key
+                    )
+                );
+            }
+            $params[$key] = $value;
+        }
+
+        return $params;
+    }
+
+    /**
+     * Prepares the messages parameter for the API request.
+     *
+     * @since n.e.x.t
+     *
+     * @param list<Message> $messages The messages to prepare.
+     * @param string|null $systemInstruction An optional system instruction to prepend to the messages.
+     * @return list<array<string, mixed>> The prepared messages parameter.
+     */
+    protected function prepareMessagesParam(array $messages, ?string $systemInstruction = null): array
+    {
+        $messagesParam = array_map(
+            function (Message $message): array {
+                // Special case: Function response.
+                $messageParts = $message->getParts();
+                if (count($messageParts) === 1 && $messageParts[0]->getType()->isFunctionResponse()) {
+                    $functionResponse = $messageParts[0]->getFunctionResponse();
+                    if (!$functionResponse) {
+                        // This should be impossible due to class internals, but still needs to be checked.
+                        throw new RuntimeException(
+                            'The function response typed message part must contain a function response.'
+                        );
+                    }
+                    return [
+                        'role' => 'tool',
+                        'content' => json_encode($functionResponse->getResponse()),
+                        'tool_call_id' => $functionResponse->getId(),
+                    ];
+                }
+                return [
+                    'role' => $this->getMessageRoleString($message->getRole()),
+                    'content' => array_values(array_filter(array_map(
+                        function (MessagePart $part): ?array {
+                            return $this->getMessagePartContentData($part);
+                        },
```

Comment:
We don't need a middleman function
```suggestion
                        [$this, 'getMessagePartContentData],
```